Modules/Core/Common/include/itkMath.h

```diff
@@ -762,6 +762,54 @@ GreatestPrimeFactor(unsigned long n);
 ITKCommon_EXPORT unsigned long long
 GreatestPrimeFactor(unsigned long long n);
 
+// C++11 does not guarantee that assert can be used in constexpr
+// functions. This is a work-around for GCC 4.8, 4.9. Originating
+// from Andrzej's C++ blog:
+//  https://akrzemi1.wordpress.com/2017/05/18/asserts-in-constexpr-functions/
+#if defined NDEBUG
+#  define ITK_X_ASSERT(CHECK) void(0)
+#else
+#  define ITK_X_ASSERT(CHECK) ((CHECK) ? void(0) : [] { assert(!#CHECK); }())
+#endif
+
+namespace Experimental
+{
+/**  Returns `a * b`. Numeric overflow triggers a compilation error in
+ * "constexpr context" and a debug assert failure at run-time.
+ */
+constexpr std::uintmax_t
+UnsignedProduct(const std::uintmax_t a, const std::uintmax_t b) ITK_NOEXCEPT
+{
+  // Note that numeric overflow is not "undefined behavior", for unsigned numbers.
+  // This function checks if the result of a*b is mathematically correct.
+  return (a == 0) || (b == 0) || ((((a * b) / a) == b) && (((a * b) / b) == a))
+           ? (a * b)
+           : (ITK_X_ASSERT(!"UnsignedProduct overflow!"), 0);
+}
+
+
+/** Calculates base ^ exponent. Numeric overflow triggers a compilation error in
+ * "constexpr context" and a debug assert failure at run-time. Otherwise equivalent to
+ * C++11 `static_cast<std::uintmax_t>(std::pow(base, exponent))`
+ *
+ * \note `UnsignedPower(0, 0)` is not supported, as zero to the power of zero does
+ * have no agreed-upon value: https://en.wikipedia.org/wiki/Zero_to_the_power_of_zero
+ */
+constexpr std::uintmax_t
+UnsignedPower(const std::uintmax_t base, const std::uintmax_t exponent) ITK_NOEXCEPT
+{
+  // Uses recursive function calls because C++11 does not support other ways of
+  // iterations for a constexpr function.
+  return (exponent == 0)
+           ? (ITK_X_ASSERT(base > 0), 1)
+           : (exponent == 1)
+               ? base
+               : UnsignedProduct(UnsignedPower(base, exponent / 2), UnsignedPower(base, (exponent + 1) / 2));
```

Comment:
Neat trick!

Reply:
Thanks, @dzenanz  Honestly, I looked at `vnl_finite_int_poly::Ntothe` (which is private, and non-`constexpr`, unfortunately): https://github.com/InsightSoftwareConsortium/ITK/blob/v5.1b01/Modules/ThirdParty/VNL/src/vxl/core/vnl/vnl_finite.h#L439